cirq-core/cirq/ops/dense_pauli_string.py

```diff
@@ -274,16 +276,24 @@ def tensor_product(self, other: 'BaseDensePauliString') -> 'DensePauliString':
         Returns:
             A dense pauli string with the concatenation of the paulis from the
             two input pauli strings, and the product of their coefficients.
+
+        Raises:
+            TypeError: If other is not of the same type than self.
         """
-        return DensePauliString(
-            coefficient=self.coefficient * other.coefficient,
-            pauli_mask=np.concatenate([self.pauli_mask, other.pauli_mask]),
-        )
+        cls = type(self)
+        if isinstance(other, cls):
+            return cls(
+                coefficient=self.coefficient * other.coefficient,
+                pauli_mask=np.concatenate([self.pauli_mask, other.pauli_mask]),
+            )
 
-    def __abs__(self):
-        return DensePauliString(coefficient=abs(self.coefficient), pauli_mask=self.pauli_mask)
+        raise TypeError(f"Tensoring allowed only with objects of type {type(self)}.")
 
-    def on(self, *qubits) -> 'cirq.PauliString':
+    def __abs__(self: T) -> T:
+        cls = type(self)
+        return cls(coefficient=abs(self.coefficient), pauli_mask=self.pauli_mask)
```

Comment:
nit: Use `type(self)(...)` instead of `cls(...)` here and elsewhere, where `cls` is used only once

Reply:
Ok